lib/src/main/kotlin/starknet/provider/HttpService.kt

```diff
@@ -0,0 +1,79 @@
+package starknet.provider
+
+import kotlinx.serialization.DeserializationStrategy
+import kotlinx.serialization.json.Json
+import okhttp3.Call
+import okhttp3.Callback
+import okhttp3.MediaType.Companion.toMediaType
+import okhttp3.OkHttpClient
+import okhttp3.RequestBody.Companion.toRequestBody
+import starknet.data.types.Response
+import java.io.IOException
+import java.util.concurrent.CompletableFuture
+
+val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()
+
+class HttpService(private val url: String, private val method: String) : Service {
+    private val client = OkHttpClient()
+
+    private fun buildHttpRequest(payload: String): okhttp3.Request {
+        val requestBody = payload.toRequestBody(JSON_MEDIA_TYPE)
+
+        return okhttp3
+            .Request
+            .Builder()
+            .url(url)
+            .method(method, requestBody)
+            .build()
+    }
+
+    private fun<T: Response> processHttpResponse(response: okhttp3.Response, deserializer: DeserializationStrategy<T>): T {
+        val responseBody = response.body
+
+        if (response.isSuccessful) {
+            if (responseBody != null) {
+                return Json.decodeFromString(deserializer, responseBody.toString())
+            } else {
+                // TODO: Create custom exception
+                throw java.lang.Exception("Empty response body")
+            }
+        } else {
+            val code = response.code
+            val text = response.body?.string() ?: "unknown"
+
+            // TODO: Create custom exception
+            throw java.lang.Exception("Invalid response, code: $code, response: $text")
+        }
+    }
+
+    override fun <T : Response> send(request: Request<T>): T {
```

Comment:
`Request` depends on `HttpService` and `HttpService` depends on `Request`. I think we can just accept payload as a string here and the Request should handle deserialization. This way we have one way dependency.

Reply:
Request depends only on Service interface, which abstracts details of HttpService.

Reply:
I was also wondering whether there should be two kinds of requests for gateway and jsonrpc2.0 calls

Reply:
>Request depends only on Service interface, which abstracts details of HttpService.

Request takes in Service in constructor, Service takes Request as a parameter. This is a cycle that could be simplified.

>I was also wondering whether there should be two kinds of requests for gateway and jsonrpc2.0 calls

I think the request model should be the same for them, or at least they should have a common interface. I guess we could get rid of service and just create specific requests if you prefer that.